apps/ui/lib/components/LoopSelector/index.tsx

```diff
@@ -0,0 +1,197 @@
+import React from 'react';
+import _ from 'lodash';
+import { useDispatch, useSelector } from 'react-redux';
+import type { Contract, LoopContract } from 'autumndb';
+import styled from 'styled-components';
+import { useHistory } from 'react-router-dom';
+import { actionCreators, selectors } from '../../store';
+import { useSetup } from '../SetupProvider';
+import Select, { SelectProps } from 'react-select';
+
+const StyledSelect = styled(Select)`
+	margin-left: 16px;
+`;
+
+// A drop-down component for selecting a loop
+export const LoopSelector = React.memo(() => {
+	const loops = useSelector(selectors.getLoops());
+	const history = useHistory();
+	const stateActiveLoop = useSelector(selectors.getActiveLoop());
+	const channels = useSelector(selectors.getChannels());
+	const { sdk } = useSetup()!;
+	const dispatch = useDispatch();
+	const [products, setProducts] = React.useState<Contract[]>([]);
+	const [activeLoop, setActiveLoop] = React.useState<LoopContract | null>(null);
+	const [activeProduct, setActiveProduct] = React.useState<Contract | null>(
+		null,
+	);
+
+	React.useEffect(() => {
+		for (const channel of channels) {
+			for (const loop of loops) {
+				if (
+					loop.id === channel.data.target ||
+					loop.slug === channel.data.target
+				) {
+					return setActiveLoop(loop);
+				}
+			}
+		}
+		setActiveLoop(null);
+		if (stateActiveLoop) {
+			dispatch(actionCreators.setActiveLoop(null));
+		}
+	}, [channels]);
+
+	React.useEffect(() => {
+		for (const channel of channels) {
+			for (const product of products) {
+				if (
+					product.id === channel.data.target ||
+					product.slug === channel.data.target
+				) {
+					if (!activeLoop) {
+						const linkedLoop = product?.links?.['is used by'][0]!;
+						if (
+							linkedLoop.id === channel.data.target ||
+							linkedLoop.slug === channel.data.target
+						) {
+							setActiveLoop(linkedLoop);
+						}
+					}
+					return setActiveProduct(product);
+				}
+			}
+		}
+		return setActiveProduct(null);
+	}, [channels]);
+
+	React.useEffect(() => {
+		if (activeLoop) {
+			sdk
+				.query({
+					type: 'object',
+					properties: {
+						type: {
+							const: 'repository@1.0.0',
+						},
+					},
+					$$links: {
+						'is used by': {
+							type: 'object',
+							properties: {
+								id: {
+									const: activeLoop.id,
+								},
+							},
+						},
+					},
+				})
+				.then((result) => {
+					setProducts(result);
+				})
```

Comment:
Plain promises to avoid self executing async function? 😄

Reply:
Yep pretty much! I'm not sure about the pros/cons of using this instead of an IIFE and async/await - what are your thoughts?

Reply:
I don't think there's much practical difference, I just avoid plain promises to stick with async/await coding style.